### PR TITLE
[runtime][Hexagon] AOTExecutor implementation for C Codegen

### PR DESCRIPTION
**Pull request by mehrdadh**
This is a follow up PR to https://github.com/apache/tvm/pull/10283 which adds hexagon implementation of AOTExecutor.

cc @kparzysz-quic @areusch

### CODE REVIEW

**Review by kparzysz-quic — COMMENTED** [review]
Some initial comments.

**Comment by areusch** [comment]
@kparzysz-quic some of these are from the underlying PRs #10283 and #10282 , i'll attempt to address them there and respond here when done.

File: src/runtime/hexagon/hexagon/hexagon_device_api_v2.cc
Comment (areusch) [reply src/runtime/hexagon/hexagon/hexagon_device_api_v2.cc:86]: on AOT we currently require device_type == kDLCPU. this hack seems consistent with the previous approach to patch the CPU device with Hexagon. we could add a comment here to explain. 

imo we should take up the issue of what device type to actually use here when we address the compilation path for Hexagon code.

File: include/tvm/runtime/metadata.h
Comment (areusch) [reply include/tvm/runtime/metadata.h]: in #10283, I moved the `#ifdef` to exclude any c++ syntax and test_cpp_aot now proves that this file compiles under C. it is my intent that this file can be compiled without the need for C++.

File: include/tvm/runtime/metadata.h
Comment (areusch) [reply include/tvm/runtime/metadata.h]: let's discuss above

File: include/tvm/runtime/metadata.h
Comment (areusch) [reply include/tvm/runtime/metadata.h]: i'll address this in #10283

File: include/tvm/runtime/metadata_base.h
Comment (areusch) [reply include/tvm/runtime/metadata_base.h]: i'll address this in #10283

File: include/tvm/runtime/metadata_base.h
Comment (areusch) [reply include/tvm/runtime/metadata_base.h]: they should be used in `src/runtime/metadata.cc`.

File: src/relay/backend/aot_executor_codegen.cc
Comment (areusch) [reply src/relay/backend/aot_executor_codegen.cc]: i'll address this in #10283

File: python/tvm/contrib/hexagon/hexagon.py
Comment (mehrdadh) [reply python/tvm/contrib/hexagon/hexagon.py:279]: that's right. Will fix this before changing PR to ready for review.

**Comment by mehrdadh** [comment]
@kparzysz-quic PTAL

File: src/runtime/hexagon/hexagon/hexagon_device_api_v2.cc
Comment (areusch) [inline src/runtime/hexagon/hexagon/hexagon_device_api_v2.cc:92]:
```suggestion
  bool is_valid_device = 
      (TVMDeviceExtType(dev.device_type) == kDLHexagon) ||
      (DLDeviceType(dev.device_type) == kDLCPU);
  CHECK(is_valid_device) << "dev.device_type: " << dev.device_type;
```

File: python/tvm/contrib/hexagon/hexagon.py
Comment (areusch) [inline python/tvm/contrib/hexagon/hexagon.py:304]: just curious, why do we register a PackedFunc to return a constant to python?

**Review by areusch — COMMENTED** [review]
overall LGTM, a minor nit and one question. could address these in a follow-on. i defer to @kparzysz-quic for final approval

File: python/tvm/contrib/hexagon/hexagon.py
Comment (mehrdadh) [reply python/tvm/contrib/hexagon/hexagon.py]: done.

**Review by kparzysz-quic — COMMENTED** [review]
A couple of comments.  Looks good in general.

File: tests/python/contrib/test_hexagon/test_launcher.py
Comment (mehrdadh) [reply tests/python/contrib/test_hexagon/test_launcher.py:242]: makes sense. I will fix this in a follow up PR if CI stays green. I have another PR which fixes few things to be able to run the tests on hardware nightly

File: tests/python/contrib/test_hexagon/test_launcher.py
Comment (mehrdadh) [reply tests/python/contrib/test_hexagon/test_launcher.py:59]: Yeah we still have that issue. I think we should try to fix it before merging your PR on adding the simulator, otherwise it would become flaky and people complain. Have you seen this issue with running on simulator?

Comment (kparzysz-quic) [reply tests/python/contrib/test_hexagon/test_launcher.py:59]: I haven't seen it outside of docker.  I don't know how to fix it, other than adding a sleep that will wait long enough for the socket to become available again...  Using different port numbers is guaranteed to work, but it requires paying attention when writing tests---it would be easy to make a mistake.

**Review by kparzysz-quic — APPROVED** [review]

**Pull request merged**
